src/ralph/models/selector.py

```diff
@@ -0,0 +1,141 @@
+"""Model selector definition"""
+
+import operator
+from collections import Counter
+from dataclasses import dataclass
+from importlib import import_module
+from itertools import chain
+from typing import Any, Callable, Union
+
+from pydantic import BaseModel
+
+from ralph.utils import get_dict_value_from_path
+
+
+@dataclass(frozen=True)
+class ModelField:
+    """Represents a model field."""
+
+    path: tuple[str]
+
+    def __init__(self, path: str):
+        object.__setattr__(self, "path", tuple(path.split(">")))
+
+
+@dataclass(frozen=True)
+class Rule:
+    """Represents a rule used for selection."""
+
+    field: ModelField
+    value: Union[ModelField, Any]
+    operation: Callable[[Any, Any], bool] = operator.eq
+
+    @staticmethod
+    def get_field_value(field, event, known_fields):
+        """Retrieves the value of a rule field.
+
+        Args:
+            field (ModelField or Any): The rule field to retrieve its value.
+            known_fields (dict): A mapping of already retrieved fields of the event.
+        """
+
+        if not isinstance(field, ModelField):
+            return field
+        if field.path in known_fields:
+            return known_fields[field.path]
+        value = get_dict_value_from_path(event, field.path)
+        known_fields[field.path] = value
+        return value
+
+    def check(self, event, known_fields):
+        """Checks if event matches the rule.
+
+        Args:
+            event (dict): The Event to check.
+            known_fields (dict): A mapping of already retrieved fields of the event.
+        """
+
+        field = self.get_field_value(self.field, event, known_fields)
+        value = self.get_field_value(self.value, event, known_fields)
+        return self.operation(field, value)
+
+
+class ModelSelector:
+    """Selects the matching model for a given event.
+
+    Attributes:
+        model_rules (dict): Stores the list of rules for each model.
+        decision_tree (dict): Stores the rule checking order for model selection.
+    """
+
+    model_rules: dict[BaseModel, list[Rule]] = {}
+
+    def __init__(self, module="ralph.models.edx"):
+        """Instantiate ModelSelector."""
+
+        import_module(module)
+        self.decision_tree = self.get_decision_tree(self.model_rules)
+
+    @classmethod
+    def add(cls, rules: list[Rule]):
+        """Event model decorator used to add event models to the model_rules list."""
+
+        def __inner__(model):
+            """Add model and rules for a given event."""
+
+            cls.model_rules[model] = rules
+            return model
+
+        return __inner__
```

Comment:
I like this new approach! Thanks) Indeed the previous method made the rule association implicit and might introduce bugs.
But I still load the edx module while instantiating to avoid circular imports. `server_event depends on Rule, Rule depends on model_selector, model_selector depends on server_event`.

Reply:
Okay. that's perfectly fine this way.